core/src/main/java/org/apache/calcite/rex/RexExecutorImpl.java

```diff
@@ -129,22 +129,20 @@ public static RexExecutable getExecutable(RexBuilder rexBuilder, List<RexNode> e
    */
   @Override public void reduce(RexBuilder rexBuilder, List<RexNode> constExps,
       List<RexNode> reducedValues) {
-    String code;
     try {
-      code = compile(rexBuilder, constExps, (list, index, storageType) -> {
+      String code = compile(rexBuilder, constExps, (list, index, storageType) -> {
         throw new UnsupportedOperationException();
       });
+
+      final RexExecutable executable = new RexExecutable(code, constExps);
+      executable.setDataContext(dataContext);
+      executable.reduce(rexBuilder, constExps, reducedValues);
     } catch (RuntimeException ex) {
       // Give up on reduction and return expressions unchanged.
       // This effectively moves the error from compile time to runtime.
       // We could give a warning here if there was a mechanism for warnings.
       reducedValues.addAll(constExps);
```

Comment:
Nit: I think your proposal will work fine as of today, but perhaps it'll be safer to do:
```
reducedValues.clear();
reducedValues.addAll(constExps);
```
Just in case in the future we might end up with an exception that leaves `reducedValues` in an intermediate state with some values filled. WDYT?

Reply:
let me make this change

Reply:
I have also added an assertion that on entry the list is empty.